File: 03-stream-processing/solution-proposal/python/exercise2/streaming_job_processing_logic.py
from pyspark.sql import DataFrame, functions as F
from pyspark.sql.types import StructType, StructField, StringType

user_visit_schema = StructType([
    StructField("os", StringType()),
    StructField("visitedPage", StringType())
])


def process_input_data(input_data_frame: DataFrame) -> DataFrame:
    return input_data_frame.select(
        F.col("key").cast("string"),
        F.from_json(F.col("value").cast("string"), user_visit_schema).alias("visit"),
        F.col("value").cast("string")
    ).filter("visit.os IN ('macOS', 'Windows', 'Linux')")\
        .filter("visit.visitedPage != ''") \
    .select(
        F.col("key"), F.col("value")
    )

File: 03-stream-processing/solution-proposal/scala/src/main/scala/com/becomedataengineer/exercise2/StreamingJobProcessingLogic.scala
package com.becomedataengineer.exercise2

import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.sql.{DataFrame, functions}

object StreamingJobProcessingLogic {

  def processInputData(inputDataFrame: DataFrame): DataFrame = {
    val schemaForFiltering = StructType(Array(
      StructField("os", StringType),
      StructField("visitedPage", StringType)
    ))

    import inputDataFrame.sparkSession.implicits._
    inputDataFrame
      .select(
        $"key".cast("string"),
        functions.from_json($"value".cast("string"), schemaForFiltering).as("visit"),
        $"value".cast("string")
      )
      .filter("visit.os IN ('macOS', 'Windows', 'Linux')")
      .filter("visit.visitedPage != ''")
      .select($"key", $"value")
  }

}
